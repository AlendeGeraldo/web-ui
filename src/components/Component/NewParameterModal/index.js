/**
 * Component responsible for:
 * - Structuring the advanced new component parameter layout
 * - Add parameter
 */
import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Form, Input, message, Modal, Select, Switch } from 'antd';
import { addParam } from '../../../store/actions/componentActions';

const { Option } = Select;
const { TextArea } = Input;

const NewParameterModal = (props) => {
  const { details, form, visible, onCancel, onAddParam } = props;
  const { uuid, parameters } = details;
  const { getFieldDecorator, getFieldsError } = form;

  /**
   * Function used to check if form has errors
   * @param {string[]} fieldsError
   */
  const hasErrors = (fieldsError) => {
    return Object.keys(fieldsError).some((field) => fieldsError[field]);
  };

  /**
   * Function to handle modal cancel
   */
  const handleCancel = () => {
    form.resetFields();
    onCancel();
  };

  /**
   * Function to handle form submit
   * @param {Event} e
   */
  const handleSubmit = (e) => {
    e.preventDefault();
    form.validateFields((err, values) => {
      if (!err) {
        const newParameter = {
          name: values.name,
          type: values.type,
          required: values.required,
          default: values.defaultValue,
          details: values.details,
        };

        onAddParam(uuid, parameters, newParameter).then((response) => {
          if (response) {
            message.success(
              `Par창metro ${newParameter.name} adicionado com sucesso`
            );
            handleCancel();
          }
        });
      }
    });
  };

  return (
    <Modal
      visible={visible}
      title='Novo Par창metro'
      okText='Criar'
      cancelText='Cancelar'
      onCancel={handleCancel}
      onOk={handleSubmit}
      okButtonProps={{ disabled: hasErrors(getFieldsError()) }}
    >
      <Form layout='vertical'>
        <Form.Item label='Nome: '>
          {getFieldDecorator('name', {
            rules: [{ required: true, message: ' ' }],
          })(<Input />)}
        </Form.Item>
        <Form.Item label='Tipo: '>
          {getFieldDecorator('type', {
            rules: [{ required: true, message: ' ' }],
          })(
            <Select>
              <Option value='float'>Float</Option>
              <Option value='int'>Int</Option>
              <Option value='string'>String</Option>
            </Select>
          )}
        </Form.Item>
        <Form.Item label='Valor padr찾o: '>
          {getFieldDecorator('defaultValue')(<Input />)}
        </Form.Item>
        <Form.Item label='Detalhes: '>
          {getFieldDecorator('details')(<TextArea />)}
        </Form.Item>
        <Form.Item label='Obrigat처rio: '>
          {getFieldDecorator('required', {
            valuePropName: 'checked',
            initialValue: true,
          })(<Switch />)}
        </Form.Item>
      </Form>
    </Modal>
  );
};

NewParameterModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

const mapStateToProps = (state) => {
  return {
    details: state.component.details,
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onAddParam: (id, parameters, newParameter) => {
      return dispatch(addParam(id, parameters, newParameter));
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Form.create({ name: 'new-parameter-modal' })(NewParameterModal));
