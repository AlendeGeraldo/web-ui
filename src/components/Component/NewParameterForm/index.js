/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable react/destructuring-assignment */
import React from 'react';

import PropTypes from 'prop-types';

import { Button, Form, Input, Select, Switch } from 'antd';

import NewParameterModal from '../NewParameterModal';

const { Option } = Select;

// eslint-disable-next-line
class NewParameterForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modalIsVisible: false,
    };
  }

  componentDidUpdate() {
    const { result, onUpdateComponentParamsReset } = this.props;
    const { modalIsVisible } = this.state;

    if (result) {
      onUpdateComponentParamsReset();
      if (modalIsVisible) {
        this.hideModal();
      } else {
        this.props.form.resetFields();
      }
    }
  }

  saveFormRef = (formRef) => {
    this.formRef = formRef;
  };

  showModal = () => {
    this.setState({ modalIsVisible: true });
  };

  hideModal = () => {
    const { form } = this.formRef.props;

    this.setState({ modalIsVisible: false });

    form.resetFields();
  };

  render() {
    const { onSubmit } = this.props;
    const { getFieldDecorator } = this.props.form;
    const { modalIsVisible } = this.state;

    const handleSubmit = (e) => {
      e.preventDefault();
      this.props.form.validateFields((err, values) => {
        if (!err) {
          onSubmit(values);
        }
      });
    };

    const handleCreateAdvancedParameter = () => {
      const { form } = this.formRef.props;
      form.validateFields(async (err, values) => {
        if (err) {
          return;
        }
        onSubmit(values);
      });
    };

    return (
      <Form onSubmit={handleSubmit}>
        <Form.Item style={{ marginBottom: 0, maxHeight: '32px' }}>
          <Form.Item
            style={{ display: 'inline-block', width: 'calc(50% - 30px)' }}
          >
            {getFieldDecorator('name', {
              rules: [{ required: true, message: ' ' }],
            })(<Input placeholder='Nome' />)}
          </Form.Item>
          <span
            style={{
              display: 'inline-block',
              width: '24px',
              textAlign: 'center',
            }}
          >
            {' '}
          </span>
          <Form.Item
            style={{ display: 'inline-block', width: 'calc(50% - 30px)' }}
          >
            {getFieldDecorator('type', {
              rules: [{ required: true, message: ' ' }],
            })(
              <Select placeholder='Tipo'>
                <Option value='float'>Float</Option>
                <Option value='int'>Int</Option>
                <Option value='string'>String</Option>
              </Select>
            )}
          </Form.Item>
        </Form.Item>
        <Form.Item style={{ marginBottom: 0 }}>
          <NewParameterModal
            wrappedComponentRef={this.saveFormRef}
            visible={modalIsVisible}
            onCancel={this.hideModal}
            onCreate={handleCreateAdvancedParameter}
          />
          <a href='#' onClick={this.showModal}>
            Avançado
          </a>
        </Form.Item>
        <Form.Item>
          {getFieldDecorator('required', {
            valuePropName: 'checked',
            initialValue: true,
          })(<Switch />)}
          <span> Obrigatório</span>
        </Form.Item>
        <Button type='primary' htmlType='submit' className='login-form-button'>
          Adicionar parâmetro
        </Button>
      </Form>
    );
  }
}

NewParameterForm.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

export default Form.create({ name: 'new_parameter' })(NewParameterForm);
