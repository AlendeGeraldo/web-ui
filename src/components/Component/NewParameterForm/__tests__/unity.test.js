import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Button, Form, Modal } from 'antd';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import NewParameterModal from '../../NewParameterModal';
import NewParameterForm from '..';

const component = {
  uuid: '1',
  createdAt: '2019-11-05T17:16:53.000Z',
  name: 'AutoML',
  parameters: [
    { name: 'b', type: 'int', required: true },
    { name: 'a', type: 'int', required: false },
  ],
};

const initialState = {
  component: {
    details: component,
  },
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

describe('NewParameterForm component', () => {
  it('is expected render without crashing', () => {
    shallow(
      <Provider store={store}>
        <NewParameterForm />
      </Provider>
    );
  });

  it('is expected render html correctly', () => {
    const wrapper = shallow(
      <Provider store={store}>
        <NewParameterForm />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render self and subcomponents', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewParameterForm />
      </Provider>
    );

    expect(wrapper.find('Form(NewParameterForm)').exists()).toBeTruthy();
    expect(wrapper.find(NewParameterForm).exists()).toBeTruthy();
    expect(wrapper.find(NewParameterModal).exists()).toBeTruthy();
  });

  it('submit should be called', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewParameterForm />
      </Provider>
    );

    const button = wrapper.find(Button);
    expect(button.simulate('click')).toBeTruthy();
  });
});
