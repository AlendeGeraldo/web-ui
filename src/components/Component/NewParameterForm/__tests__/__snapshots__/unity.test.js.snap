// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewParameterForm component is expected render html correctly 1`] = `
<NewParameterForm
  details={
    Object {
      "createdAt": "2019-11-05T17:16:53.000Z",
      "name": "AutoML",
      "parameters": Array [
        Object {
          "name": "b",
          "required": true,
          "type": "int",
        },
        Object {
          "name": "a",
          "required": false,
          "type": "int",
        },
      ],
      "uuid": "1",
    }
  }
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
  onAddParam={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
/>
`;
